types/src/blob.rs

```diff
@@ -135,6 +138,133 @@ impl Blob {
+
+        for _ in 1..shares_needed {
+            let Some(share) = shares.next() else {
+                return Err(Error::MissingShares);
+            };
+            if share
+                .info_byte()
+                .map(|info| info.version() != share_version)
+                .unwrap_or(true)
+            {
+                return Err(Error::BlobSharesMetadataMismatch("share version".into()));
+            }
+            if share.namespace() != namespace {
+                return Err(Error::BlobSharesMetadataMismatch("namespace".into()));
+            }
+            if share.sequence_length().is_some() {
+                return Err(Error::UnexpectedSequenceStart);
+            }
+            data.extend_from_slice(share.payload().expect("non parity"));
+        }
+
+        // remove padding
+        data.truncate(blob_len as usize);
+
+        Self::new(namespace, data)
+    }
+
+    /// Reconstructs all blobs from shares.
+    ///
+    /// This function will seek shares that indicate start of next blob (with
+    /// [`Share::sequence_length`]) and pass them to [`Blob::reconstruct`].
+    /// It will automatically ignore all shares that are within reserved namespaces,
+    /// e.g. it is completely fine to pass whole [`ExtendedDataSquare`] to this
+    /// function and get all blobs in the block.
+    ///
+    /// # Errors
+    ///
+    /// This function propagates any errors from [`Blob::reconstruct`].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use celestia_types::Blob;
+    /// # use celestia_types::nmt::Namespace;
+    /// # let namespace1 = Namespace::new_v0(&[1, 2, 3, 4, 5]).expect("Invalid namespace");
+    /// # let namespace2 = Namespace::new_v0(&[2, 3, 4, 5, 6]).expect("Invalid namespace");
+    ///
+    /// let blobs = vec![
+    ///     Blob::new(namespace1, b"foo".to_vec()).unwrap(),
+    ///     Blob::new(namespace2, b"bar".to_vec()).unwrap(),
+    /// ];
+    /// let shares: Vec<_> = blobs.iter().flat_map(|blob| blob.to_shares().unwrap()).collect();
+    ///
+    /// let reconstructed = Blob::reconstruct_all(&shares).unwrap();
+    ///
+    /// assert_eq!(blobs, reconstructed);
+    /// ```
+    ///
+    /// [`ExtendedDataSquare`]: crate::ExtendedDataSquare
+    pub fn reconstruct_all<'a, I>(shares: I) -> Result<Vec<Self>>
+    where
+        I: IntoIterator<Item = &'a Share>,
+    {
+        let mut shares = shares
+            .into_iter()
+            .filter(|shr| !shr.namespace().is_reserved());
+        let mut blobs = Vec::with_capacity(2);
+
+        loop {
+            let mut blob = {
+                // find next share from blobs namespace that is sequence start
+                let Some(start) = shares.find(|&shr| shr.sequence_length().is_some()) else {
+                    break;
+                };
+                iter::once(start).chain(&mut shares)
+            };
+            blobs.push(Blob::reconstruct(&mut blob)?);
```

Comment:
should this error out if shares iterator ends while we are in the middle of some blob?
I'm wondering if it shouldn't be a valid usecase where user passes eg. 3 rows and we get all blobs that they contain, without failing if 3rd row only has part of some next blob